libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance/SpecTranslate/Conway.hs

```diff
@@ -1327,3 +1328,29 @@ instance
       <$> toSpecRep lsUTxOState
       <*> toSpecRep (utxosGovState lsUTxOState ^. proposalsGovStateL)
       <*> toSpecRep lsCertState
+
+instance
+  ( EraPParams era
+  , ConwayEraGov era
+  , SpecTranslate ctx (PParamsHKD Identity era)
+  , SpecRep (PParamsHKD Identity era) ~ Agda.PParams
+  , SpecTranslate ctx (PParamsHKD StrictMaybe era)
+  , SpecRep (PParamsHKD StrictMaybe era) ~ Agda.PParamsUpdate
+  , Inject ctx [GovActionState era]
+  , ToExpr (PParamsHKD StrictMaybe era)
+  , SpecRep (TxOut era) ~ Agda.TxOut
+  , SpecTranslate ctx (TxOut era)
+  ) =>
+  SpecTranslate ctx (NewEpochState era)
+  where
+  type SpecRep (NewEpochState era) = Agda.NewEpochState
+
+  toSpecRep (NewEpochState {..}) =
+    Agda.MkNewEpochState
+      <$> toSpecRep nesEL
+      <*> toSpecRep nesEs
+
+instance SpecTranslate ctx (ConwayNewEpochPredFailure era) where
+  type SpecRep (ConwayNewEpochPredFailure era) = OpaqueErrorString
+
+  toSpecRep _ = pure $ OpaqueErrorString "TODO"
```

Comment:
I was running into some trouble deriving `ToExpr` instances for `ConwayNewEpochPredFailure`, hence the `TODO`. I also tried to generate `ToExpr` instances for `ShelleyNewEpochPredFailure` (out of curiosity) and I ran into the same problem as with the one for `Conway`: the constraints regarding the protocol versions and min, max versions could not be satisfied (e.g.: `MinVersion <= ProtVerHigh era`). I have not been able to figure out the reason why yet.

Reply:
So out of curiosity, if you define these : 
`instance ToExpr (RewardUpdate c)`
`instance ToExpr (ConwayNewEpochPredFailure era)` 

and then `  toSpecRep e = pure . OpaqueErrorString .show $ toExpr e`  instead of the "TODO",

where is it failing?  When you run the test? It's failing atm anyway, but is it failing differently if these are defined? 
What does it mean to "generate ToExpr instances "?

Reply:
For `MinVersion <= ProtVerHigh era`, I have seen this one in the SRP PR, and I found it to be part of the `Era` class. I had to go through many hoops to finally satisfy GHC and still not need to use this explicitly. In my case it was because I was using `era` in place of `EraCrypto era` in a few places. But also, because I had been using some constraints in some instances that did not require those. So if the constraints you have placed above are not due to the compiler explicitly asking for them, as in they are copied, then it might help to add them one by one and following the compiler errors. This is very counter-intuitive to relying on ghc to correctly point out `redundant constraints`, which it fails to some times.

Reply:
@teodanciu if I add those instances, it won't even compile and GHC will complain like so: 
```
testlib/Test/Cardano/Ledger/Conway/TreeDiff.hs:272:3: error: [GHC-64725]
    • Cannot satisfy: MinVersion <= ProtVerHigh (EraCrypto era)
```
But fortunately it seems like Aniket already figured out how to satisfy these. 

@aniketd Yup, I was looking at the `Era` class as well and that is why this whole thing seemed so strange to me because I assumed that GHC would already know what the `era` parameter is by the time we are deriving the `ToExpr` instance for `ConwayNewEpochPredFailure` (and thus it would know that those `MinVersion <= ProtVerHigh era` constraints hold). It seems like that's not the case though. I'm happy to learn that you ran into the same thing and managed to resolve it. I'll try to replicate what you've done and see how it goes. Thanks! :raised_hands:

Reply:
This is now resolved. In my case the problem was the fact that I tried to parameterise the `ToExpr` and `SpecTranslate` instances for `ConwayNewEpochPredFailure` while everything else related was already concretised as `ConwayEra StandardCrypto`.